### PR TITLE
Refactor stat

### PR DESCRIPTION
**Pull request by Unengine**
## Refactors

 - `StatMap`, `StatMapEx` is removed. `DecimalStat` does the same thing with them.
 - Made `DecimalStat` serializable considering backward compatibility of `StatMap` and `StatMapEx`.
 - `StatMap` **(Completely different from class above!)** is added to handle total stat of `CharacterBase` and `ItemUsable`. 
 Removes duplicate code between `CharacterStats` which determines total stat of character and `StatsMap` which determines total stat of `ItemUsable`.
 - Added new stat `ArmorPenetration` and `Thorn`.
 - Every stat is now calculated in `decimal`. But value of stat is converted to `int` on actual use.

### CODE REVIEW

**Review by ipdae — CHANGES_REQUESTED** [review]
target branch will be released someday, this code should not be merged until the fix InvalidBlockStateRootHashException

File: Lib9c/Model/Stat/CharacterStats.cs [hunk 1]
```diff
@@ -249,85 +186,67 @@ public CharacterStats SetStats(int level, bool updateImmediate = true)
...
-                            statMap.CDMG));
-                    }
+        /// <summary>
+        /// Set stats based on runes.
+        /// </summary>
+        /// <param name="value"></param>
+        /// <param name="updateImmediate"></param>
+        /// <returns></returns>
+        public CharacterStats SetRunes(IEnumerable<StatModifier> value, bool updateImmediate = true)
+        {
```
Comment (boscohyun) [inline Lib9c/Model/Stat/CharacterStats.cs:230]: I wonder if this new specification is backward compatible.

**Review by boscohyun — COMMENTED** [review]
Tomorrow I will review `Nekoyume.Model.Stat.DecimalStat` and the rest.

File: .Lib9c.DevExtensions.Tests/Action/CreateOrReplaceAvatarTest.cs [hunk 2]
```diff
@@ -506,9 +506,9 @@ public CreateOrReplaceAvatarTest()
                 foreach (var statOption in statOptions)
                 {
                     Assert.Contains(
-                        equipment.StatsMap.GetAdditionalStats(),
-                        stat => stat.StatType == statOption.Key &&
-                                stat.AdditionalValue == statOption.Value);
+                        equipment.StatsMap.GetAdditionalStats(true),
```
[the remark this answers is not in the document]
Comment (Unengine) [reply .Lib9c.DevExtensions.Tests/Action/CreateOrReplaceAvatarTest.cs:509]: This is not intended. I will fix it later!

File: .Lib9c.Tests/Action/HackAndSlash15Test.cs [hunk 3]
```diff
@@ -1032,14 +1033,14 @@ public void CheckRewardItems(bool backward, int worldId, int stageId)
...
-        public void CheckCrystalRandomSkillState(bool clear, bool skillStateExist, bool hasCrystalSkill)
+        public void CheckCrystalRandomSkillState(bool forceClear, bool skillStateExist, bool hasCrystalSkill)
         {
             const int worldId = 1;
             const int stageId = 5;
             const int clearedStageId = 4;
             var previousAvatarState = _initialState.GetAvatarStateV2(_avatarAddress);
             previousAvatarState.actionPoint = 999999;
-            previousAvatarState.level = clear ? 400 : 3;
+            previousAvatarState.level = forceClear ? 400 : 1;
```
[the remark this answers is not in the document]
Comment (Unengine) [reply .Lib9c.Tests/Action/HackAndSlash15Test.cs:1043]: To make sure the player defeated in the battle.

File: Lib9c.DevExtensions/Action/CreateOrReplaceAvatar.cs [hunk 4]
```diff
@@ -497,10 +497,10 @@ public override IAccountStateDelta Execute(IActionContext context)
                     optionSheet,
                     null,
                     skillSheet);
-                var additionalOptionStats = equipment.StatsMap.GetAdditionalStats().ToArray();
+                var additionalOptionStats = equipment.StatsMap.GetAdditionalStats(true).ToArray();
```
[the remark this answers is not in the document]
Comment (Unengine) [reply Lib9c.DevExtensions/Action/CreateOrReplaceAvatar.cs:500]: This is not intended, too.

File: Lib9c/Battle/CPHelper.cs [hunk 5]
```diff
@@ -140,6 +140,10 @@ public static decimal GetStatCP(StatType statType, decimal statValue, int charac
                     return GetCPOfDRR(statValue, characterLevel);
                 case StatType.CDMG:
                     return GetCPOfCDMG(statValue, characterLevel);
+                case StatType.ArmorPenetration:
+                    return GetCPOfArmorPenetration(statValue);
+                case StatType.DamageReflection:
+                    return GetCPOfDamageReflection(statValue);
```
[the remark this answers is not in the document]
Comment (Unengine) [reply Lib9c/Battle/CPHelper.cs:146]: I've already checked when implementing `DamageReduction`! It's ok to add new `StatType`.

File: Lib9c/Model/Character/ArenaCharacter.cs [hunk 6]
```diff
@@ -218,7 +233,6 @@ private static CharacterSheet.Row CharacterRow(int characterId, ArenaSimulatorSh
 
             stats.SetOption(options);
             stats.IncreaseHpForArena();
-            stats.EqualizeCurrentHPWithHP();
```
[the remark this answers is not in the document]
Comment (Unengine) [reply Lib9c/Model/Character/ArenaCharacter.cs:221]: Yes. `CurrentHP` is moved to `CharacterBase` and `ArenaCharacter`.

File: Lib9c/Model/Character/CharacterBase.cs [hunk 7]
```diff
@@ -92,7 +100,6 @@ public int CurrentHP
             defElementType = RowData.ElementalType;
             RunSpeed = RowData.RunSpeed;
             attackRange = RowData.AttackRange;
-            CurrentHP = HP;
```
[the remark this answers is not in the document]
Comment (Unengine) [reply Lib9c/Model/Character/CharacterBase.cs:95]: Same as above.

File: Lib9c/Model/Stat/CharacterStats.cs (hunk 1 above)
Comment (Unengine) [reply Lib9c/Model/Stat/CharacterStats.cs:230]: It's not. I should fix it when handling state mismatches.

**Review by ipdae — APPROVED** [review]

**Pull request merged**